quickwit-core/src/indexing/index.rs

```diff
@@ -0,0 +1,123 @@
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use futures::try_join;
+use quickwit_metastore::IndexUri;
+use quickwit_metastore::Metastore;
+use quickwit_metastore::{MetastoreUriResolver, SplitState};
+use quickwit_storage::StorageUriResolver;
+use tokio::sync::mpsc::channel;
+use tracing::warn;
+
+use crate::index::garbage_collect;
+use crate::indexing::document_retriever::DocumentSource;
+use crate::indexing::split_finalizer::finalize_split;
+use crate::indexing::statistics::StatisticsCollector;
+use crate::indexing::{document_indexer::index_documents, split::Split};
+
+const SPLIT_CHANNEL_SIZE: usize = 30;
+
+/// A struct to bundle index cli args
+/// TODO: remove when there is a better structure
+#[derive(Debug, Clone)]
+pub struct IndexDataParams {
+    pub index_uri: PathBuf,
+    pub input_uri: Option<PathBuf>,
+    pub temp_dir: PathBuf,
+    pub num_threads: usize,
+    pub heap_size: u64,
+    pub overwrite: bool,
+}
+
+/// The entry point for the index command.
+/// It reads a new-line deleimited json documents, add them to the index while building
+/// and publishing splits metastore.
+pub async fn index_data(params: IndexDataParams) -> anyhow::Result<()> {
+    let index_uri = params.index_uri.to_string_lossy().to_string();
+    let metastore = MetastoreUriResolver::default().resolve(&index_uri)?;
+    let storage_resolver = Arc::new(StorageUriResolver::default());
+
+    if params.overwrite {
+        reset_index(index_uri, storage_resolver.clone(), metastore.clone()).await?;
+    }
+
+    if params.input_uri.is_none() {
+        println!("Please enter your new line delimited json documents.");
+    }
+
+    let document_retriever = Box::new(DocumentSource::create(&params.input_uri).await?);
+    let (statistic_collector, statistic_sender) = StatisticsCollector::start_collection();
+    let (split_sender, split_receiver) = channel::<Split>(SPLIT_CHANNEL_SIZE);
+    try_join!(
+        index_documents(
+            &params,
+            metastore,
+            storage_resolver,
+            document_retriever,
+            split_sender,
+            statistic_sender.clone()
+        ),
+        finalize_split(split_receiver, statistic_sender.clone()),
```

Comment:
Since publishing is part of finalizing, splits will be published as they are produced. Ideally, we'd like to publish all the splits at once when indexing is done (see this [comment](https://github.com/quickwit-inc/quickwit/issues/3#issuecomment-843839740)).

We can address that in a follow-up PR.

Reply:
ok